src/QMCHamiltonians/MomentumEstimator.cpp

```diff
@@ -172,115 +147,271 @@ void MomentumEstimator::setParticlePropertyList(PropertySetType& plist
-  std::vector<int> qk(0);
-  mappedQtonofK.resize(numqtwists,qk);
-  compQ.resize(numqtwists);
-  RealType qn(4.0*M_PI*M_PI*std::pow(Lattice.Volume,-2.0/3.0));
-  mappedQnorms.resize(numqtwists,qn*0.5/RealType(M));
-  if (twist[0]==0)
-    mappedQnorms[kgrid]=qn/RealType(M);
-  if (twist[1]==0)
-    mappedQnorms[3*kgrid+1]=qn/RealType(M);
-  if (twist[2]==0)
-    mappedQnorms[5*kgrid+2]=qn/RealType(M);
-  Q.resize(numqtwists);
-  for (int i=-kgrid; i<(kgrid+1); i++)
+  PosType kmaxs(0);
+  kmaxs[0]=kmax0;
+  kmaxs[1]=kmax1;
+  kmaxs[2]=kmax2;
+  bool sphere_only=true;
+  bool directional_only=true;
+  RealType sum_kmaxs=kmaxs[0]+kmaxs[1]+kmaxs[2];
+  RealType sphere_kmax;
+  if (kmax<1.0e-10 && sum_kmaxs<1.0e-10)
   {
-    PosType kpt;
-    kpt[0]=i-twist[0];
-    kpt[1]=i-twist[1];
-    kpt[2]=i-twist[2];
-    kpt=Lattice.k_cart(kpt);
-    Q[i+kgrid]=std::abs(kpt[0]);
-    Q[i+kgrid+(2*kgrid+1)]=std::abs(kpt[1]);
-    Q[i+kgrid+(4*kgrid+2)]=std::abs(kpt[2]);
+    // default: kmax = 2 x k_F of polarized non-interacting electron system
+    kmax = 2.0*std::pow(6.0*M_PI*M_PI*elns.getTotalNum()/elns.Lattice.Volume,1.0/3);
+    kgrid = int(kmax/min_Length)+1;
+    sphere_kmax=kmax;
+    sphere_only=true;
+    directional_only=false;
+  }
+  else
+  {
+    sphere_kmax=kmax;
+    if (sum_kmaxs>0.0 && kmax>0.0)
+    {
+      directional_only=false;
+      sphere_only=false;
+    }
+    else
+    {
+      if (kmax<1.0e-10)
+        sphere_only=false;
+      if (sum_kmaxs<1.0e-10)
+        directional_only=false;
+    }
+    for (int i=0; i<OHMMS_DIM; i++)
+    {
+      if (kmaxs[i]>kmax)
+        kmax=kmaxs[i];
+    }
+    kgrid = int(kmax/min_Length)+1;
   }
-  app_log()<<" Using all k-space points with (nx^2+ny^2+nz^2)^0.5 < "<< kgrid <<" for Momentum Distribution."<< std::endl;
-  app_log()<<"  My twist is:"<<twist[0]<<"  "<<twist[1]<<"  "<<twist[2]<< std::endl;
-  int indx(0);
-  int kgrid_squared=kgrid*kgrid;
+  int kgrid_squared[OHMMS_DIM];
+  for (int i=0; i<OHMMS_DIM; i++)
+    kgrid_squared[i]=int(std::round(kmaxs[i]*kmaxs[i]/vec_length[i]/vec_length[i]));
+  RealType kmax_squared=sphere_kmax*sphere_kmax;
+  std::vector<int> kcount0;
+  std::vector<int> kcount1;
+  std::vector<int> kcount2;
+  kcount0.resize(2*kgrid+1,0);
+  kcount1.resize(2*kgrid+1,0);
+  kcount2.resize(2*kgrid+1,0);
   for (int i=-kgrid; i<(kgrid+1); i++)
   {
     for (int j=-kgrid; j<(kgrid+1); j++)
     {
       for (int k=-kgrid; k<(kgrid+1); k++)
       {
-        if (i*i+j*j+k*k<=kgrid_squared) //if (std::sqrt(i*i+j*j+k*k)<=kgrid)
+        PosType kpt;
+        kpt[0]=i-twist[0];
+        kpt[1]=j-twist[1];
+        kpt[2]=k-twist[2];
+        //convert to Cartesian: note that 2Pi is multiplied
+        kpt=Lattice.k_cart(kpt);
+        bool not_recorded=true;
+        // This collects the k-points within the parallelepiped (if not in sphere_only mode)
+        if (!sphere_only && i*i<=kgrid_squared[0] && j*j<=kgrid_squared[1] && k*k<=kgrid_squared[2])
```

Comment:
Need to include the twist

Reply:
The twist is included in the actual k-point, and therefore, for the parallelepiped the grid will be a twist shifted parallelepiped when done this way. This is what Jaron and I thought would be best in general, and also, it is convenient for our next simulations. To this end, this way the values kmax0, kmax1, and kmax2 are not absolute limiting values, but e.g. kmax0+|twist0| is possible.

Reply:
The sphere kmax checking includes twist so it need to be consistent.
I think the checking should be (i-twist[0])*(i-twist[0])<=kgrid_squared[0]

Reply:
In the (i-twist[0]) case the kgrid_squared does need to be float, which is what you referred to in your previous comment.

Reply:
We will probably revert this change in a subsequent PR.  It is better to have a uniform grid size on all twist groups and this is better for postprocessing also.

Reply:
Apart from postprocessing convenience, is there any other reason for physics?